lib/jobs/generate-tag.js

```diff
@@ -0,0 +1,119 @@
+// Copyright 2015, EMC, Inc.
+
+'use strict';
+
+var di = require('di');
+
+module.exports = generateTagJobFactory;
+di.annotate(generateTagJobFactory, new di.Provide('Job.Catalog.GenerateTag'));
+di.annotate(generateTagJobFactory, new di.Inject(
+    'Job.Base',
+    'Services.Waterline',
+    'Protocol.Events',
+    'JobUtils.CatalogSearchHelpers',
+    'anchor',
+    'Logger',
+    'Util',
+    'Promise',
+    'Assert',
+    '_'
+));
+function generateTagJobFactory(
+    BaseJob,
+    waterline,
+    eventsProtocol,
+    catalogSearch,
+    anchor,
+    Logger,
+    util,
+    Promise,
+    assert,
+    _
+) {
+
+    var logger = Logger.initialize(generateTagJobFactory);
+
+    /**
+     *
+     * @param {Object} [options]
+     * @constructor
+     */
+    function GenerateTagJob(options, context, taskId) {
+        GenerateTagJob.super_.call(this, logger, options, context, taskId);
+
+        this.nodeId = context.target || options.nodeId;
+        assert.isMongoId(this.nodeId, 'context.target || options.nodeId');
+    }
+
+    util.inherits(GenerateTagJob, BaseJob);
+
+    var RULE_DELIMITER = /\./g;
+
+    /**
+     * @memberOf GenerateTagJob
+     * @returns {Promise}
+     */
+    GenerateTagJob.prototype._run = function run() {
+        var self = this;
+
+        waterline.tags.find({}).then(function (tags) {
+            var catalogTypes = _(tags)
+                .pluck('rules')
+                .flattenDeep()
+                .pluck('path')
+                .map(function (path) {
+                    return _.compact(path.split(RULE_DELIMITER))[0];
+                })
+                .uniq()
+                .value();
+            return [ Promise.map(catalogTypes, function(type) {
+                return waterline.catalogs.findMostRecent({
+                    node: self.nodeId,
+                    source: type
+                });
+            }), tags ];
+        }).spread(function(catalogs, tags) {
+            catalogs = _(catalogs)
+                .flattenDeep()
+                .compact()
+                .transform(function (catalogs, catalog) {
+                    catalogs[catalog.source] = catalog.data;
+                }, {})
+                .value();
+
+            var matches = _.filter( matchTags(catalogs, tags), function(match) {
+                return match.tag.rules.length && !match.errors.length;
+            });
+
+            var tagNames = _(matches).pluck('tag.name').value();
+            if( !_.isEmpty(tagNames)) {
+                return waterline.nodes.addTags(tagNames);
+            }
+        }).then(function() {
+            self._done();
+        }).catch(function(err) {
+            self._done(err);
+        });
+    };
+
+    function matchTags(catalogs, tags) {
```

Comment:
@yyscamper yeah I agree I think formalizing on something like JSONPath is a good idea to start work on. Mongodb doesn't support `.` characters within keys because the mongo query language allows for nested key searches, for example:

```
find({
    'key.subkey.subsubkey': <value>
})
```

would find a document like:

```
{
    "key": {
        "subkey": {
            "subsubkey": <value>
        }
    }
}
```